porch/pkg/apiserver/webhooks.go

```diff
@@ -0,0 +1,326 @@
+	if err != nil {
+		return nil, err
+	}
+
+	serverCertPEMBytes := serverCertPEM.Bytes()
+	err = WriteFile("/etc/webhook/certs/tls.crt", serverCertPEMBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	serverPrivKeyPEMBytes := serverPrivKeyPEM.Bytes()
+	err = WriteFile("/etc/webhook/certs/tls.key", serverPrivKeyPEMBytes)
+	if err != nil {
+		return nil, err
+	}
+	return caPEM.Bytes(), nil
+}
+
+// WriteFile writes data in the file at the given path
+func WriteFile(filepath string, sCert []byte) error {
+	fmt.Println(string(sCert))
+	f, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = f.Write(sCert)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func createValidatingWebhook(ctx context.Context, caCert []byte) error {
+	var (
+		webhookNamespace, _  = os.LookupEnv("WEBHOOK_NAMESPACE")
+		validationCfgName, _ = os.LookupEnv("VALIDATE_CONFIG")
+		webhookService, _    = os.LookupEnv("WEBHOOK_SERVICE")
+	)
+	config := ctrl.GetConfigOrDie()
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		panic("failed to set go -client")
+	}
+
+	path := "/validate-deletion"
+	fail := admissionregistrationv1.Fail
+	none := admissionregistrationv1.SideEffectClassNone
+
+	caCert = bytes.TrimSuffix(caCert, []byte("\n"))
+
+	validateConfig := &admissionregistrationv1.ValidatingWebhookConfiguration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: validationCfgName,
+		},
+		Webhooks: []admissionregistrationv1.ValidatingWebhook{{
+			Name: "packagerevdeletion.google.com",
+			ClientConfig: admissionregistrationv1.WebhookClientConfig{
+				CABundle: caCert, // CA bundle created earlier
+				Service: &admissionregistrationv1.ServiceReference{
+					Name:      webhookService,
+					Namespace: webhookNamespace,
+					Path:      &path,
+				},
+			},
+			Rules: []admissionregistrationv1.RuleWithOperations{{Operations: []admissionregistrationv1.OperationType{
+				admissionregistrationv1.Delete},
+				Rule: admissionregistrationv1.Rule{
+					APIGroups:   []string{"porch.kpt.dev"},
+					APIVersions: []string{"v1alpha1"},
+					Resources:   []string{"packagerevisions"},
+				},
+			}},
+			AdmissionReviewVersions: []string{"v1", "v1beta1"},
+			SideEffects:             &none,
+			FailurePolicy:           &fail,
+		}},
+	}
+
+	if _, err := kubeClient.AdmissionregistrationV1().ValidatingWebhookConfigurations().Create(ctx, validateConfig,
+		metav1.CreateOptions{}); err != nil {
+	}
+
+	return nil
+}
+
+var logger = log.New(os.Stdout, "http: ", log.LstdFlags)
+
+func runWebhookServer(certFile, keyFile string) {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("Starting webhook server")
+	http.HandleFunc("/validate-deletion", validateDeletion)
+	server := http.Server{
+		Addr: fmt.Sprintf(":%d", 443),
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+		},
+	}
+
+	if err := server.ListenAndServeTLS("", ""); err != nil {
```

Comment:
This tries to start a server on the same port that is already in use by the aggregated apiserver. We need to either find a way to expose additional endpoints on the apiserver, use a different port, or do this in a separate pod. I think the second option might be the best one.